pinephone/vibrator/FFDevice.cpp

```diff
@@ -0,0 +1,135 @@
+/* SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (C) 2019 The Android Open-Source Project
+ * Copyright (C) 2021 GloDroid project
+ */
+
+#include "vibrator-impl/FFDevice.h"
+
+#include <android-base/logging.h>
+
+typedef bool (*dirfunc_t)(const char *path, void *param);
+static bool recursive_dir_observer(const char *path, dirfunc_t func,
+                                   void *param);
+
+namespace aidl {
+namespace android {
+namespace hardware {
+namespace vibrator {
+
+std::unique_ptr<FFDevice> FFDevice::create(const char *input_path) {
+    std::string event_path;
+    auto observe_func = [](const char *path, void *param) -> bool {
+        std::string *event_path = static_cast<std::string *>(param);
+        std::string p = path;
+        if (p.find("event") != std::string::npos) {
+            *event_path = p;
+            return true;
+        }
+        return false;
+    };
+    if (recursive_dir_observer(input_path, observe_func, &event_path)) {
+        event_path = "/dev/input/" + event_path;
+        auto device = std::unique_ptr<FFDevice>(new FFDevice());
```

Comment:
Consider replacing this with `auto device = std::make_unique<FFDevice>()`. We aren't worried about exception safety there, I believe, but it is better to stick to the common best practices.